File: sov.php
<?php
/*
###############################################################################
#
#    This file is part of CRVSTVLA
#
#    CRVSTVLA is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    CRVSTVLA is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with CRVSTVLA; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
###############################################################################
*/

include_once 'i18n.inc.php';


$declinatio = N_("si non declinatione neque sermonem transferre");
$declinatio_est = T_($declinatio) != $declinatio;

$alinea = "<br/>\n";

$nominM = array(
    N_("Aemilius"),
    N_("Eupalamus"),
    N_("Publius"),
    N_("Aulus"),
    N_("Marcus"));

$nominF = array(
    N_("Aemilia"),
    N_("Iulia"),
    N_("Lucia"),
    N_("Lucretia"),
    N_("Sempronia"));

$accusM = array(
   N_("Aemilium"),
   N_("Eupalamum"),
   N_("Publium"),
   N_("Aulum"),
   N_("Marcum"));

$accusF = array(
    N_("Aemiliam"),
    N_("Iuliam"),
    N_("Luciam"),
    N_("Lucretiam"),
    N_("Semproniam"));

$nomina = array_merge($nominF, $nominM);
$accus = array_merge($accusF,$accusM);
$nominADaccus=array();
$accusADnomin=array();

function accusatiuus_est($n){
    global $accus;
    return in_array($n, $accus, true);
}
             
for($i=0; $i < count($nomina); $i++){
    $nominADaccus[$nomina[$i]] = $accus[$i];
    $accusADnomin[$accus[$i]] = $nomina[$i];
}

function gallice($s) {
    // in tabulam uertere sententiam :
    global $sujet, $objet, $accusADnomin, $declinatio_est;
    $s = preg_replace("/\W$/", "", $s);
    $coll = explode(" ", $s);
    foreach($coll as $uer) { 
        //echo '*'.$uer.'*'.$alinea;
        if ($uer == "amat")
            $verbe = T_("aime");
        elseif (accusatiuus_est($uer))
            $objet = $uer;
        else $sujet = $uer;
    }
    if ($declinatio_est){
        $sujet = T_($sujet);
        $objet = T_($objet);
    } else {
        $objet = $accusADnomin[$objet];
    }
    return "$sujet $verbe $objet."; 
}

function sorsColl($c) {
    if (count($c) == 1) return $c[0];
   return $c[mt_rand(0, count($c)-1)];   
}

// décider du sexe du sujet
$sexus = array('m','f');
if (sorsColl($sexus) == 'm') {
    $subiectus = sorsColl($nominM);
    $lemma_obiecti = sorsColl($nominF);
} else {
    $subiectus = sorsColl($nominF);
    $lemma_obiecti = sorsColl($nominM);
}
$obiectus = $nominADaccus[$lemma_obiecti];

function sent() {
    global $subiectus, $obiectus, $alinea;
    $c = array ($subiectus, $obiectus, 'amat');
    shuffle($c);
    return implode (" ", $c) . '.';
}

function IIPropositiones()
{
   // Potius quam nomen quaerere inter multa,
   // II sententias proponere
    global $declinatio_est;
    global $subiectus, $lemma_obiecti, $obiectus;
    global $accusADnomin, $nominADaccus;
    if ($declinatio_est){
        $s1 = T_($subiectus);
        $o1 = T_($obiectus);
        $s2 = T_($accusADnomin[$obiectus]);
        $o2 = T_($nominADaccus[$subiectus]);
    } else { // non est declinatio
        $s1 = $subiectus;
        $o1 = $lemma_obiecti;
        $s2 = $lemma_obiecti;
        $o2 = $subiectus;
    }
    $p = array(
        "$s1 " . T_("aime") . " $o1.",
        "$s2 " . T_("aime") . " $o2.");
    shuffle ($p);
    return $p; 
}

session_start(); 
echo "<html>\n"
	."<head>\n"
	."<title>CRVSTVLA - SOV</title>\n"
	."<link rel=\"stylesheet\" href=\"crustula.css\" type=\"text/css\">\n"
    ."<link rel=\"icon\" href=\"favicon.ico\">\n";
//include "css.inc";
include "meta.inc.php";   
echo "</head>\n"
	."<body>\n"
	."<p class=\"titre\">\n"
	."QVIS QVEM AMAT ?\n"
	."</p>\n";

if (isset($_REQUEST['sententia']))
    $priorsent = $_POST["sententia"]; 
if (isset($priorsent)){
   	echo "prior sententia : ".$priorsent.$alinea;
   	$gallice = gallice($priorsent);
   	echo T_("gallice :")." ".$gallice.$alinea;
   	$resp = $_POST["resp"];
   	//echo "<br>sujet : $sujet, objet : $objet.<br>";
   	$recte = $resp == $gallice;
   	if ($recte) echo "<div class=\"juste\">RECTE !</div>"; 
   	else echo "<div class=\"faux\"> Errauisti. Respondisti $resp</div>";
   	include 'session.php.html';
} else {
  	$_SESSION['prius'] = 0;
  	$_SESSION['consec'] = 0;
}
echo "</p>\n" 
	."<p class=\"question\">\n";
$sententia = sent();
echo $sententia;
echo "</p>\n"
	."<form method=\"post\">\n"
	."<p class=\"question\">\n";
$trans = IIPropositiones();
foreach ($trans as $p) 
{
    echo "<input type=\"submit\" name= \"resp\" value=\"$p\" class=\"question\"><br /><br />\n";
}
echo "<input type=\"hidden\" name=\"sententia\" value=\"$sententia\">";
?>
</form>
<p class="redi"><a href='./'>reditus</a> <a href="gpl.html" target="gpl">Licentia GPL</a></p>
</body>
</html>

File: latin/crustula/crustula/sov.py
###############################################################################
#
#    This file is part of CRVSTVLA
#
#    CRVSTVLA is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    CRVSTVLA is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with CRVSTVLA; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
###############################################################################
from django.shortcuts import render
from .utils.i18n import *
from .utils.recte import *
from crustula.models import *
import random

try:
    vocab = Sov.objects.get(name="sov")
    

    M = list(vocab.ov_set.filter(genre="m"))
    F = list(vocab.ov_set.filter(genre="f"))
    acc = [(o.accLatine, o) for o in vocab.ov_set.all()] # tous les accusatifs latins
    nom = [(o.nomLatine, o) for o in vocab.ov_set.all()] # tous les nominatifs latins

except:
    print("Warning: the database should contain data in the table crustula_ov, referenced by the name 'sov' in the table crustula_sov ... something went wrong!")
    pass

def accusatiuus_est(uer):
    """
    Vérifie si uer est un accusatif
    @return un Ov ayant uer pour accusatif, sinon None 
    """
    for latine, o in acc:
        if latine == uer:
            return o
    return

def nominatiuus_est(uer):
    """
    Vérifie si uer est un nominatif
    @return un Ov ayant uer pour nominatif, sinon None 
    """
    for latine, o in nom:
        if latine == uer:
            return o
    return

def f_gallice(s):
    ## in tabulam uertere sententiam :
    while s.endswith("."): s = s[:-1]
    coll = s.split(" ")
    for uer in coll:
        if uer == "amat":
            verbe = _("aime")
            continue
        a = accusatiuus_est(uer)
        if a:
            objet = a
            continue
        n = nominatiuus_est(uer)
        if n:
            sujet = n
    return f"{sujet.nomGallice} {verbe} {objet.accGallice}."

def alea_jacta_est():
    """
    décider du sexe du sujet
    @return une paire d'OV : un sujet et un objet de genre mf ou fm
    """
    sexus =('m','f')
    if (random.choice(sexus) == 'm'):
        return random.choice(M), random.choice(F)
    else:
        return random.choice(F), random.choice(M)

def IIPropositiones():
    """
    fabrique de façon aléatoire le groupe sujet, verbe objet
    @return une phrase, et deux propositions
    """
    subiectus, obiectus = alea_jacta_est()
    s1 = subiectus.nomGallice
    o1 = obiectus.accGallice
    s2 = obiectus.nomGallice
    o2 = subiectus.accGallice
    p = [
        f"{s1} {_('aime')} {o1}.",
        f"{s2} {_('aime')} {o2}.",
    ]
    random.shuffle(p)
    c = [subiectus.nomLatine, obiectus.accLatine, 'amat']
    random.shuffle(c)
    phrase = " ".join(c) + "."
    return phrase, p[0], p[1]


def index(request):
    preferred_language(request)
    priorsent = request.POST.get("sententia","")
    gallice = ""
    resp=""
    recte = False
    prius = request.session.get("prius", 0)
    request.session["prius"] = prius
    consec = request.session.get("consec", 0)
    request.session["consec"] = consec
    if priorsent:
        gallice = f_gallice(priorsent)
        resp = request.POST.get("resp","")
        recte = resp == gallice
    compte_points(request, recte)
    sententia, r1, r2 = IIPropositiones()
    return render(request,'crustula/sov.html', context={
        "priorsent": priorsent,
        "gallice": gallice,
        "lingue": _("gallice"),
        'prius': request.session['prius'],
        'consec': request.session['consec'],
        "resp": resp,
        "question": sententia,
        "recte": recte,
        "r1": r1,
        "r2": r2,
    })

